### PR TITLE
fix(k8sprocessor): fix data race in k8sprocessor and some some tests

### PR DESCRIPTION
**Pull request by pmalek-sumo**
> NOTE:
> There is a problematic aspect of those tests which is in detail described in the comment in `owner_test.go`. One can verify the problem by running e.g. `while go test -v -count 100 -race -run Test_Owner ./kube/...; do ; done` and observing that the tests sometimes fail.

The primary concern of this PR is below mentioned data race.

At the same time add some unit tests which would run the changed code.

```
==================
WARNING: DATA RACE
Write at 0x00c00127a420 by goroutine 44:
  runtime.mapdelete_faststr()
      runtime/map_faststr.go:300 +0x0
  github.com/open-telemetry/opentelemetry-collector-contrib/processor/k8sprocessor/kube.(*WatchClient).deleteLoop()
      github.com/open-telemetry/opentelemetry-collector-contrib/processor/k8sprocessor@v0.41.0/kube/client.go:203 +0x564
  github.com/open-telemetry/opentelemetry-collector-contrib/processor/k8sprocessor/kube.New·dwrap·1()
      github.com/open-telemetry/opentelemetry-collector-contrib/processor/k8sprocessor@v0.41.0/kube/client.go:78 +0x58

Previous read at 0x00c00127a420 by goroutine 249:
  github.com/open-telemetry/opentelemetry-collector-contrib/processor/k8sprocessor/kube.(*WatchClient).handlePodUpdate()
      github.com/open-telemetry/opentelemetry-collector-contrib/processor/k8sprocessor@v0.41.0/kube/client.go:162 +0x2e4
  github.com/open-telemetry/opentelemetry-collector-contrib/processor/k8sprocessor/kube.(*WatchClient).handlePodUpdate-fm()
      github.com/open-telemetry/opentelemetry-collector-contrib/processor/k8sprocessor@v0.41.0/kube/client.go:154 +0x6d
  k8s.io/client-go/tools/cache.ResourceEventHandlerFuncs.OnUpdate()
      k8s.io/client-go@v0.22.4/tools/cache/controller.go:238 +0x8b
  k8s.io/client-go/tools/cache.(*ResourceEventHandlerFuncs).OnUpdate()
      <autogenerated>:1 +0x3d
  k8s.io/client-go/tools/cache.(*processorListener).run.func1()
      k8s.io/client-go@v0.22.4/tools/cache/shared_informer.go:775 +0x27a
  k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1()
      k8s.io/apimachinery@v0.22.4/pkg/util/wait/wait.go:155 +0x81
  k8s.io/apimachinery/pkg/util/wait.BackoffUntil()
      k8s.io/apimachinery@v0.22.4/pkg/util/wait/wait.go:156 +0xce
  k8s.io/apimachinery/pkg/util/wait.JitterUntil()
      k8s.io/apimachinery@v0.22.4/pkg/util/wait/wait.go:133 +0x104
  k8s.io/apimachinery/pkg/util/wait.Until()
      k8s.io/apimachinery@v0.22.4/pkg/util/wait/wait.go:90 +0x78
  k8s.io/client-go/tools/cache.(*processorListener).run()
      k8s.io/client-go@v0.22.4/tools/cache/shared_informer.go:771 +0x18
  k8s.io/client-go/tools/cache.(*processorListener).run-fm()
      k8s.io/client-go@v0.22.4/tools/cache/shared_informer.go:765 +0x39
  k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1()
      k8s.io/apimachinery@v0.22.4/pkg/util/wait/wait.go:73 +0x7e

Goroutine 44 (running) created at:
  github.com/open-telemetry/opentelemetry-collector-contrib/processor/k8sprocessor/kube.New()
      github.com/open-telemetry/opentelemetry-collector-contrib/processor/k8sprocessor@v0.41.0/kube/client.go:78 +0x498
  github.com/open-telemetry/opentelemetry-collector-contrib/processor/k8sprocessor.(*kubernetesprocessor).initKubeClient()
      github.com/open-telemetry/opentelemetry-collector-contrib/processor/k8sprocessor@v0.41.0/processor.go:50 +0x2e5
  github.com/open-telemetry/opentelemetry-collector-contrib/processor/k8sprocessor.createKubernetesProcessor()
      github.com/open-telemetry/opentelemetry-collector-contrib/processor/k8sprocessor@v0.41.0/factory.go:165 +0x23a
  github.com/open-telemetry/opentelemetry-collector-contrib/processor/k8sprocessor.createLogsProcessorWithOptions()
      github.com/open-telemetry/opentelemetry-collector-contrib/processor/k8sprocessor@v0.41.0/factory.go:134 +0xc6
  github.com/open-telemetry/opentelemetry-collector-contrib/processor/k8sprocessor.createLogsProcessor()
      github.com/open-telemetry/opentelemetry-collector-contrib/processor/k8sprocessor@v0.41.0/factory.go:73 +0xd0
  go.opentelemetry.io/collector/processor/processorhelper.(*factory).CreateLogsProcessor()
      go.opentelemetry.io/collector@v0.41.0/processor/processorhelper/factory.go:133 +0x168
  go.opentelemetry.io/collector/service/internal/builder.(*pipelinesBuilder).buildPipeline()
      go.opentelemetry.io/collector@v0.41.0/service/internal/builder/pipelines_builder.go:203 +0x937
  go.opentelemetry.io/collector/service/internal/builder.BuildPipelines()
      go.opentelemetry.io/collector@v0.41.0/service/internal/builder/pipelines_builder.go:105 +0x25c
  go.opentelemetry.io/collector/service.newService()
      go.opentelemetry.io/collector@v0.41.0/service/service.go:73 +0xadb
  go.opentelemetry.io/collector/service.(*Collector).setupConfigurationComponents()
      go.opentelemetry.io/collector@v0.41.0/service/collector.go:192 +0x572
  go.opentelemetry.io/collector/service.(*Collector).Run()
      go.opentelemetry.io/collector@v0.41.0/service/collector.go:231 +0x3d8
  go.opentelemetry.io/collector/service.NewCommand.func1()
      go.opentelemetry.io/collector@v0.41.0/service/command.go:40 +0x452
  github.com/spf13/cobra.(*Command).execute()
      github.com/spf13/cobra@v1.2.1/command.go:856 +0xa7d
  github.com/spf13/cobra.(*Command).ExecuteC()
      github.com/spf13/cobra@v1.2.1/command.go:974 +0x5da
  github.com/spf13/cobra.(*Command).Execute()
      github.com/spf13/cobra@v1.2.1/command.go:902 +0x5c
  main.runInteractive()
      github.com/SumoLogic/sumologic-otel-collector/main.go:34 +0x60
  main.run()
      github.com/SumoLogic/sumologic-otel-collector/main_others.go:11 +0x237
  main.main()
      github.com/SumoLogic/sumologic-otel-collector/main.go:27 +0x49
  runtime.main()
      runtime/proc.go:255 +0x226
  github.com/cilium/ebpf/link.init()
      github.com/cilium/ebpf@v0.6.2/link/syscalls.go:50 +0x2a8
  github.com/cilium/ebpf/link.init()
      github.com/cilium/ebpf@v0.6.2/link/syscalls.go:29 +0x24a
  runtime.doInit()
      runtime/proc.go:6498 +0x122
  github.com/cilium/ebpf.init()
      github.com/cilium/ebpf@v0.6.2/syscalls.go:426 +0x690
  github.com/cilium/ebpf.init()
      github.com/cilium/ebpf@v0.6.2/prog.go:458 +0x630
  github.com/cilium/ebpf.init()
      github.com/cilium/ebpf@v0.6.2/syscalls.go:408 +0x5d0
  github.com/cilium/ebpf.init()
      github.com/cilium/ebpf@v0.6.2/syscalls.go:236 +0x570
  github.com/cilium/ebpf.init()
      github.com/cilium/ebpf@v0.6.2/syscalls.go:220 +0x510
  github.com/cilium/ebpf.init()
      github.com/cilium/ebpf@v0.6.2/syscalls.go:203 +0x4b0
  github.com/cilium/ebpf.init()
      github.com/cilium/ebpf@v0.6.2/syscalls.go:185 +0x44e
  runtime.doInit()
      runtime/proc.go:6498 +0x122
  github.com/cilium/ebpf/internal/btf.init()
      github.com/cilium/ebpf@v0.6.2/internal/btf/btf.go:728 +0x1bc

Goroutine 249 (running) created at:
  k8s.io/apimachinery/pkg/util/wait.(*Group).Start()
      k8s.io/apimachinery@v0.22.4/pkg/util/wait/wait.go:71 +0xde
  k8s.io/client-go/tools/cache.(*sharedProcessor).run.func1()
      k8s.io/client-go@v0.22.4/tools/cache/shared_informer.go:623 +0x1da
  k8s.io/client-go/tools/cache.(*sharedProcessor).run()
      k8s.io/client-go@v0.22.4/tools/cache/shared_informer.go:627 +0x51
  k8s.io/client-go/tools/cache.(*sharedProcessor).run-fm()
      k8s.io/client-go@v0.22.4/tools/cache/shared_informer.go:618 +0x44
  k8s.io/apimachinery/pkg/util/wait.(*Group).StartWithChannel.func1()
      k8s.io/apimachinery@v0.22.4/pkg/util/wait/wait.go:56 +0x3e
  k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1()
      k8s.io/apimachinery@v0.22.4/pkg/util/wait/wait.go:73 +0x7e
==================
```

### CODE REVIEW

**Review by pmm-sumo — APPROVED** [review]

**Comment by swiatekm-sumo** [comment]
I wonder if we shouldn't be using a thread-safe data structure to store the pod/owner data to begin with. The cache package has https://pkg.go.dev/k8s.io/client-go/tools/cache#Store and https://pkg.go.dev/k8s.io/client-go/tools/cache#Indexer, and the latter would also make storing Pod data under multiple keys easier to manage. It's easy to make small mistakes using mutexes manually for this.

This PR looks good to me, asking more generally. @pmalek-sumo @pmm-sumo ?

**Comment by pmalek-sumo** [comment]
> I wonder if we shouldn't be using a thread-safe data structure to store the pod/owner data to begin with. The cache package has https://pkg.go.dev/k8s.io/client-go/tools/cache#Store and https://pkg.go.dev/k8s.io/client-go/tools/cache#Indexer, and the latter would also make storing Pod data under multiple keys easier to manage. It's easy to make small mistakes using mutexes manually for this.
> 
> This PR looks good to me, asking more generally. @pmalek-sumo @pmm-sumo ?

In general the idea seems ok but I'd say that's not the primary concern for us (as long as the tests pass and in general this seems to work).

My main concern here is the flakiness of tests that I've added due to missing endpoint creation notifications :/

File: pkg/processor/k8sprocessor/kube/owner_test.go [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+// ...
+//
+//
+// This is most probably an issue with the way the framework is being used.
+// Minimal example with a single informer and 2 endpoints being added always deliver 2
+// object creation notification.
+//
+// Due to the above we're not adding below commented out assertions to the pipeline
+// but they might be a good indication (if uncommented) of general correctness of
+// the system.
```
Comment (pmalek-sumo) [inline pkg/processor/k8sprocessor/kube/owner_test.go:61]: 🤷

**Comment by pmalek-sumo** [comment]
Here's a working test example which mimics what's being done in ownerCache but succeeds 100% of the time (I wasn't able to make it fail even after running tests in a loop for a while)

```go
package kube

import (
	"context"
	"fmt"
	"sync/atomic"
	"testing"
	"time"

	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
	api_v1 "k8s.io/api/core/v1"
	meta_v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/apimachinery/pkg/fields"
	"k8s.io/apimachinery/pkg/labels"
	"k8s.io/client-go/informers"
	"k8s.io/client-go/kubernetes/fake"
	"k8s.io/client-go/tools/cache"
)

func Test_InformerNotifications(t *testing.T) {
	const (
		ns = "kube-system"
	)
	var (
		counter int32
		pod     = &api_v1.Pod{
			ObjectMeta: metav1.ObjectMeta{
				Name:      "my-pod",
				Namespace: ns,
				UID:       "f15f0585-a0bc-43a3-96e4-dd2eace75392",
			},
		}
		endpoints = &api_v1.Endpoints{
			ObjectMeta: metav1.ObjectMeta{
				Name:      "my-service",
				Namespace: pod.Namespace,
				UID:       "f15f0585-a0bc-43a3-96e4-dd2eace75390",
			},
			TypeMeta: metav1.TypeMeta{
				Kind: "Endpoint",
			},
			Subsets: []api_v1.EndpointSubset{
				{
					Addresses: []api_v1.EndpointAddress{
						{
							TargetRef: &api_v1.ObjectReference{
								Name:      pod.Name,
								Namespace: pod.Namespace,
								Kind:      "Pod",
								UID:       pod.UID,
							},
						},
					},
				},
			},
		}
		endpoints2 = &api_v1.Endpoints{
			ObjectMeta: metav1.ObjectMeta{
				Name:      "my-service-2",
				Namespace: pod.Namespace,
				UID:       "f15f0585-a0bc-43a3-96e4-dd2eace75389",
			},
			TypeMeta: metav1.TypeMeta{
				Kind: "Endpoint",
			},
			Subsets: []api_v1.EndpointSubset{
				{
					Addresses: []api_v1.EndpointAddress{
						{
							TargetRef: &api_v1.ObjectReference{
								Name:      pod.Name,
								Namespace: pod.Namespace,
								Kind:      "Pod",
								UID:       pod.UID,
							},
						},
					},
				},
			},
		}
	)

	c := fake.NewSimpleClientset()

	factory := informers.NewSharedInformerFactoryWithOptions(c, watchSyncPeriod,
		informers.WithNamespace(ns),
		informers.WithTweakListOptions(func(opts *meta_v1.ListOptions) {
			opts.LabelSelector = labels.Everything().String()
			opts.FieldSelector = fields.Everything().String()
		}))

	informer := factory.Core().V1().Endpoints().Informer()
	informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
		AddFunc: func(obj interface{}) {
			e, ok := obj.(*api_v1.Endpoints)
			if ok {
				atomic.AddInt32(&counter, 1)
				fmt.Println("add")
			} else {
				fmt.Printf("WTF? %#v\n", e)
			}
		},
		UpdateFunc: func(_, obj interface{}) {
			fmt.Println("update")
		},
		DeleteFunc: func(obj interface{}) {
			fmt.Println("delete")
		},
	})

	_, err := c.CoreV1().Endpoints(ns).
		Create(context.Background(), endpoints, metav1.CreateOptions{})
	require.NoError(t, err)

	_, err = c.CoreV1().Endpoints(ns).
		Create(context.Background(), endpoints2, metav1.CreateOptions{})
	require.NoError(t, err)

	_, err = c.CoreV1().Pods(ns).
		Create(context.Background(), pod, metav1.CreateOptions{})
	require.NoError(t, err)

	ch := make(chan struct{})
	defer close(ch)
	go func() {
		informer.Run(ch)
	}()

	if !assert.Eventually(t, func() bool {
		return atomic.LoadInt32(&counter) == 2
	}, 5*time.Second, 10*time.Millisecond) {
		panic("")
	}
}
```

**Pull request merged**